# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/Lowering.cpp
# Commit: 174a622bf334
# Full Hash: 174a622bf334e32f911e97ca76c36b0a80d51f94
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-11 15:10:42
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.5: Add tls dependency for WasmModD. r=lth
#   
#   We generate builtin call for Mod operation for Double types, so we need
#   to add a tls dependency. In this patch I've added it.
#   
# ==============================================================================

diff -r 49039661a807 -r 174a622bf334 js/src/jit/Lowering.cpp
--- a/js/src/jit/Lowering.cpp	Fri Sep 11 11:08:35 2020 +0000
+++ b/js/src/jit/Lowering.cpp	Fri Sep 11 11:20:20 2020 +0000
@@ -1853,6 +1853,14 @@
   lowerWasmBuiltinModI64(mod);
 }
 
+void LIRGenerator::visitWasmBuiltinModD(MWasmBuiltinModD* ins) {
+  MOZ_ASSERT(gen->compilingWasm());
+  LWasmBuiltinModD* lir = new (alloc()) LWasmBuiltinModD(
+      useRegisterAtStart(ins->lhs()), useRegisterAtStart(ins->rhs()),
+      useFixedAtStart(ins->tls(), WasmTlsReg));
+  defineReturn(lir, ins);
+}
+
 void LIRGenerator::visitMod(MMod* ins) {
   MOZ_ASSERT(ins->lhs()->type() == ins->rhs()->type());
   MOZ_ASSERT(IsNumberType(ins->type()));
@@ -1876,13 +1884,13 @@
     MOZ_ASSERT(ins->lhs()->type() == MIRType::Double);
     MOZ_ASSERT(ins->rhs()->type() == MIRType::Double);
 
-    // Ion does an unaligned ABI call and thus needs a temp register. Wasm
-    // doesn't.
-    LDefinition maybeTemp = gen->compilingWasm() ? LDefinition::BogusTemp()
-                                                 : tempFixed(CallTempReg0);
-
-    LModD* lir = new (alloc()) LModD(useRegisterAtStart(ins->lhs()),
-                                     useRegisterAtStart(ins->rhs()), maybeTemp);
+    MOZ_ASSERT(!gen->compilingWasm());
+
+    // Ion does an unaligned ABI call and thus needs a temp register.
+    // Note: useRegisterAtStart is safe here, the temp is not a FP register.
+    LModD* lir = new (alloc())
+        LModD(useRegisterAtStart(ins->lhs()), useRegisterAtStart(ins->rhs()),
+              tempFixed(CallTempReg0));
     defineReturn(lir, ins);
     return;
   }